apps/condo/domains/ticket/components/analytics/TicketChartView.tsx

```diff
@@ -35,60 +46,161 @@ const TicketChartView: React.FC<ITicketAnalyticsPageChartProps> = ({
 }) => {
     const intl = useIntl()
     const NoData = intl.formatMessage({ id: 'NoData' })
+    let legend = [], axisData = {}, tooltip = {}
+    const [chartReadyCounter, setChartReadyCounter] = useState<number>(0)
+    // Start from 1 because used as multiplier with TICKET_CHART_PAGE_SIZE
+    const [pieChartPage, setPieChartPage] = useState(1)
+    // Store pie chart refs because we need to access api for every chart component
+    const chartRefs = useRef([])
+    // Cache series result for client side paging
+    const seriesRef = useRef([])
+    const seriesCacheRef = useRef([])
+
+    if (data !== null) {
+        const mapperResult = mapperInstance.getChartConfig(viewMode, data)
+        seriesCacheRef.current = mapperResult.series
+        const pageSize = TICKET_CHART_PAGE_SIZE * pieChartPage
+        seriesRef.current = mapperResult.series.slice(0, pageSize)
+        // If component used as report widget, we should use page separation. Otherwise we load all data (for pdf page)
+        // series = seriesRef.current.slice(0, pageSize)
+        legend = mapperResult.legend
+        axisData = mapperResult.axisData
+        tooltip = mapperResult.tooltip
+    }
+
+    useEffect(() => {
+        // Clean chart refs if input data was changed
+        if (viewMode === 'pie') {
+            if (seriesRef.current.length !== chartRefs.current.length) {
+                // It means that filters were changed & total chart count may be changed too, so just reset refs & page
+                setChartReadyCounter(0)
+                setPieChartPage(1)
+            }
+            chartRefs.current = chartRefs.current.slice(0, seriesRef.current.length)
+        }
+    }, [data])
+
+    // Way to await moment when all pie chart instance were rendered (needed for client side pdf generation)
+    useLayoutEffect(() => {
+        if (viewMode === 'pie' && onChartReady !== undefined) {
```

Comment:
I'm not sure we need to use check for ```onChartReady``` here, guess we should move it on 87 line

Reply:
`useLayoutEffect` should be used only for pdf generation when pie chart view were selected. So this check in right place